routers/common/logger_contxt.go

```diff
@@ -0,0 +1,229 @@
+		fi.funcFileShort = fi.funcName
+	}
+
+	// remove package prefix. eg: "xxx.com/pkg1/pkg2.foo" => "pkg2.foo"
+	p1 = strings.LastIndexByte(fi.funcName, '/')
+	if p1 >= 0 {
+		fi.funcNameShort = fi.funcName[p1+1:]
+	} else {
+		fi.funcNameShort = fi.funcName
+	}
+
+	// remove ".func?" suffix for anonymous func
+	// usually we do not have more than 10 anonymous functions in one function, so this check is enough and fast.
+	if len(fi.funcNameShort) > 6 {
+		if fi.funcNameShort[len(fi.funcNameShort)-6:len(fi.funcNameShort)-1] == ".func" {
+			fi.funcNameShort = fi.funcNameShort[:len(fi.funcNameShort)-6]
+		}
+	}
+	return fi
+}
+
+var contextKeyLogRequestRecord interface{} = "logRequestRecord"
+
+//UpdateContextHandlerFuncInfo updates a context's func info by a real handler func `v`
+func UpdateContextHandlerFuncInfo(ctx context.Context, v interface{}, friendlyName ...string) {
+	if reqRec, ok := ctx.Value(contextKeyLogRequestRecord).(*logRequestRecord); ok {
+		var fi *logFuncInfo
+		ptr := reflect.ValueOf(v).Pointer()
+
+		funcInfoMapMu.RLock()
+		fi, ok = funcInfoMap[ptr]
+		funcInfoMapMu.RUnlock()
+
+		if !ok {
+			f := runtime.FuncForPC(ptr)
+			if f != nil {
+				fi = runtimeFuncToInfo(f)
+				if len(friendlyName) == 1 {
+					fi.funcNameShort = friendlyName[0]
+				}
+
+				funcInfoMapMu.Lock()
+				funcInfoMap[ptr] = fi
+				funcInfoMapMu.Unlock()
+			}
+		}
+
+		reqRec.funcInfoMu.Lock()
+		reqRec.funcInfo = fi
+		reqRec.funcInfoMu.Unlock()
+	}
+}
+
+// WrapContextHandler wraps a log context handler for a router handler
+func WrapContextHandler(pathSuffix string, handler http.HandlerFunc, friendlyName ...string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			if !strings.HasPrefix(req.URL.Path, pathSuffix) {
+				next.ServeHTTP(resp, req)
+				return
+			}
+			UpdateContextHandlerFuncInfo(req.Context(), handler, friendlyName...)
+			handler(resp, req)
+		})
+	}
+}
+
+//UpdateContextHandlerPanicError updates a context's error info, a panic may be recovered by other middlewares, but we still need to know that.
+func UpdateContextHandlerPanicError(ctx context.Context, err interface{}) {
+	if reqRec, ok := ctx.Value(contextKeyLogRequestRecord).(*logRequestRecord); ok {
+		reqRec.panicError = err
+	}
+}
+
+func (lh *logContextHandler) startSlowQueryDetector(threshold time.Duration) {
+	graceful.GetManager().RunWithShutdownContext(func(baseCtx context.Context) {
+		go func() {
+			// This go-routine checks all active requests every second.
+			// If a request has been running for a long time (eg: /user/events), we also print a log with "still-executing" message
+			// After the "still-executing" log is printed, the record will be removed from the map to prevent from duplicated logs in future
+			t := time.NewTicker(time.Second)
+			for {
+				select {
+				case <-baseCtx.Done():
+					return
+				case <-t.C:
```

Comment:
I don't think the ticker is the right thing here as this function could take more than a second to process

```suggestion
			for {
				select {
				case <-baseCtx.Done():
					return
				case <-time.After(time.Second):
```

Reply:
I think the ticker is the right thing. We do not care about accurate duration here. It just does the check periodically, 500ms or 1500ms are all OK. And the code runs really fast here, no problem. I did my best to make sure everything work as fast as possible.

Using NewTicker can save a memory allocation, because time.After always allocates a new Timer struct.